pkg/cloud/externalconn/impl_registry.go

```diff
@@ -16,23 +16,86 @@ import (
+var parseAndValidateFns = map[string]schemeRegistration{}
+
+// FactoryType identifies the expected parsing logic used to create an external connection.
+// The only one currently supported is SimpleURIFactory.
+type FactoryType int
+
+const (
+	SimpleURIFactory FactoryType = iota
+)
+
+var factoryFactories = map[FactoryType]func(connectionpb.ConnectionProvider) connectionParserFactory{
+	SimpleURIFactory: makeSimpleURIFactory,
+}
 
 // RegisterConnectionDetailsFromURIFactory is used by every concrete
 // implementation to register its factory method.
 func RegisterConnectionDetailsFromURIFactory(
-	providerScheme string, parseAndValidateFn connectionParserFactory,
+	providerScheme string, provider connectionpb.ConnectionProvider, factoryType FactoryType,
 ) {
-	if _, ok := parseAndValidateFns[providerScheme]; ok {
-		panic(fmt.Sprintf("parse function already registered for %s", providerScheme))
+	if existing, ok := parseAndValidateFns[providerScheme]; ok {
+		if existing.FactoryType != factoryType || existing.ConnectionProvider != provider {
+			panic(fmt.Sprintf("different parse function already registered for %s", providerScheme))
+		}
+	} else {
+		parseAndValidateFns[providerScheme] = schemeRegistration{
+			FactoryType:             factoryType,
+			ConnectionProvider:      provider,
+			connectionParserFactory: factoryFactories[factoryType](provider),
+			validations:             make(map[string]ValidationFn),
+		}
+	}
+}
+
+const defaultValidation = `default`
+
+// RegisterDefaultValidation sets the default validation for external connections using this scheme.
+func RegisterDefaultValidation(providerScheme string, validation ValidationFn) {
+	RegisterNamedValidation(providerScheme, defaultValidation, validation)
+}
+
+// RegisterNamedValidation adds a custom validation for external connections using this scheme.
+// It will be the default if no default is provided.
+func RegisterNamedValidation(providerScheme string, name string, validation ValidationFn) {
+	existing, ok := parseAndValidateFns[providerScheme]
+	if !ok {
+		panic(fmt.Sprintf("no parse function registered yet for %s", providerScheme))
+	}
+	existing.validations[name] = validation
+	if _, ok = existing.validations[defaultValidation]; !ok {
+		existing.validations[defaultValidation] = validation
+	}
+	parseAndValidateFns[providerScheme] = existing
+}
+
+func makeSimpleURIFactory(provider connectionpb.ConnectionProvider) connectionParserFactory {
+	return func(ctx context.Context, execCfg interface{}, user username.SQLUsername, uri *url.URL) (ExternalConnection, error) {
+		connDetails := connectionpb.ConnectionDetails{
+			Provider: provider,
+			Details: &connectionpb.ConnectionDetails_SimpleURI{
+				SimpleURI: &connectionpb.SimpleURI{
+					URI: uri.String(),
+				},
+			},
+		}
+		return NewExternalConnection(connDetails), nil
 	}
-	parseAndValidateFns[providerScheme] = parseAndValidateFn
 }
 
 // ExternalConnectionFromURI returns a ExternalConnection for the given URI.
```

Comment:
Can I ask you to flesh out this comment to reflect that it creates an External Connection and runs the registered validation which could include pinging the resource.